File: src/org/vsegda/data/DataItem.java
package org.vsegda.data;

import com.google.appengine.api.datastore.Key;
import org.vsegda.util.TimeUtil;

import javax.jdo.annotations.IdGeneratorStrategy;
import javax.jdo.annotations.PersistenceCapable;
import javax.jdo.annotations.Persistent;
import javax.jdo.annotations.PrimaryKey;

/**
 * @author Roman Elizarov
 */
@PersistenceCapable
public class DataItem {
	@PrimaryKey
    @Persistent(valueStrategy = IdGeneratorStrategy.IDENTITY)
    private Key key;

	@Persistent
	private long streamId;

	@Persistent
	private double value;

	@Persistent
	private long timeMillis;

    public DataItem() {}

    public DataItem(String line, long now) {
        String[] tokens = line.split(",");
        if (tokens.length < 2 || tokens.length > 3)
            throw new IllegalArgumentException("Invalid line format: " + line);
        try {
            streamId = Long.parseLong(tokens[0]);
            value = Double.parseDouble(tokens[1]);
            timeMillis = tokens.length < 3 ? now : TimeUtil.parseTime(tokens[2], now);
        } catch (NumberFormatException e) {
            throw new IllegalArgumentException("Invalid line format: " + line, e);
        }
	}

    public Key getKey() {
        return key;
    }

    public void setKey(Key key) {
        this.key = key;
    }

    public long getStreamId() {
        return streamId;
    }

    public void setStreamId(long streamId) {
        this.streamId = streamId;
    }

    public double getValue() {
		return value;
	}

	public void setValue(double value) {
		this.value = value;
	}

	public long getTimeMillis() {
		return timeMillis;
	}

	public void setTimeMillis(long timeMillis) {
		this.timeMillis = timeMillis;
	}

    public String getTime() {
        return TimeUtil.formatDateTime(timeMillis);
    }

    public String getAgo() {
        return TimeUtil.formatDateTimeDifference(getTimeMillis(), System.currentTimeMillis());
    }

    public String getFormatClass() {
        return TimeUtil.getFormatClass(getTimeMillis(), System.currentTimeMillis());
    }

    @Override
	public String toString() {
        return streamId + "," + value + "," + TimeUtil.formatDateTime(timeMillis);
	}
}

File: service/src/main/java/org/vsegda/data/DataItem.kt
package org.vsegda.data

import org.vsegda.util.*

class DataItem(
    var streamId: Long = 0L,
    var value: Double = 0.0,
    var timeMillis: Long = 0L
) {
    var itemId: Long = 0L

    // non-persistent!!!
    private var _stream: DataStream? = null

    val time: String
        get() = TimeUtil.formatDateTime(timeMillis)

    val ago: String
        get() = TimeUtil.formatDateTimeDifference(timeMillis, System.currentTimeMillis())

    val formatClass: String
        get() = TimeUtil.getFormatClass(timeMillis, System.currentTimeMillis())

    val isRecent: Boolean
        get() = timeMillis >= System.currentTimeMillis() - RECENT_TIME_INTERVAL

    var stream: DataStream
        get() = _stream ?: DataStream(streamId)
        set(stream) {
            if (streamId == 0L)
                streamId = stream.streamId
            else
                require(streamId == stream.streamId)
            _stream = stream
        }

    /**
     * Returns stream id or tag of the stream if defined.
     */
    @Suppress("MemberVisibilityCanBePrivate")
    val streamCode: String
        get() = _stream?.code ?: streamId.toString()

    constructor(line: String, now: Long) : this() {
        val tokens = line.split(",").dropLastWhile { it.isEmpty() }.toTypedArray()
        require(tokens.size in 2..3) { "Invalid line format: $line" }
        stream = DataStream()
        stream.tag = tokens[0]
        try {
            value = tokens[1].toDouble()
            timeMillis = if (tokens.size < 3) now else Math.min(TimeUtil.parseTime(tokens[2], now), now)
        } catch (e: NumberFormatException) {
            throw IllegalArgumentException("Invalid line format: " + line, e)
        }
    }

    constructor(stream: DataStream, value: Double, timeMillis: Long) :
        this(stream.streamId, value, timeMillis)

    override fun toString(): String = "$streamCode,$value,$time"

    companion object {
        val BY_TIME = compareBy<DataItem> { it.timeMillis }
    }
}
